server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java

```diff
@@ -103,54 +131,89 @@ public Leaf leaf(LeafReader reader, int[] docIdsInLeaf) throws IOException {
 
     /**
      * Load a field for {@link Synthetic}.
+     * <p>
+     * {@link SyntheticFieldLoader}s load values through objects vended
+     * by their {@link #storedFieldLoaders} and {@link #docValuesLoader}
+     * methods. Then you call {@link #write} to write the values to an
+     * {@link XContentBuilder} which also clears them.
+     * <p>
+     * This two loaders and one writer setup is specifically designed to
+     * efficiently load the {@code _source} of indices that have thousands
+     * of fields declared in the mapping but that only have values for
+     * dozens of them. It handles this in a few ways:
+     * <ul>
+     *     <li>{@link #docValuesLoader} must be called once per document
+     *         per field to load the doc values, but detects up front if
+     *         there are no doc values for that field. It's linear with
+     *         the number of fields, whether or not they have values,
+     *         but skips entirely missing fields.</li>
+     *     <li>{@link #storedFieldLoaders} are only called when we see
+     *         a stored field and are otherwise ignored. So it's fine
```

Comment:
Can you clarify what you mean by 'when we see a stored field' - I guess you mean when we find that a field is present in the current document's stored fields?

Reply:
yes. I'll push a comment for it.